File: android/app/src/main/java/com/mobileproxy/app/network/Protocol.kt
package com.mobileproxy.app.network

import org.json.JSONObject

object Protocol {
    const val REQUEST_ID_LENGTH = 36

    // Create a register message
    fun createRegisterMessage(deviceKey: String, name: String, carrier: String?, ip: String?): String {
        val deviceInfo = JSONObject().apply {
            put("name", name)
            carrier?.let { put("carrier", it) }
            ip?.let { put("ip", it) }
        }
        return JSONObject().apply {
            put("type", "register")
            put("deviceKey", deviceKey)
            put("deviceInfo", deviceInfo)
        }.toString()
    }

    // Create a pong message
    fun createPongMessage(): String {
        return JSONObject().apply { put("type", "pong") }.toString()
    }

    // Create proxy response headers message
    fun createResponseHeaders(requestId: String, statusCode: Int, headers: Map<String, String>): String {
        val headersJson = JSONObject()
        headers.forEach { (key, value) -> headersJson.put(key, value) }
        return JSONObject().apply {
            put("type", "proxy_response_headers")
            put("requestId", requestId)
            put("statusCode", statusCode)
            put("headers", headersJson)
        }.toString()
    }

    // Create proxy response end message
    fun createResponseEnd(requestId: String): String {
        return JSONObject().apply {
            put("type", "proxy_response_end")
            put("requestId", requestId)
        }.toString()
    }

    // Create proxy error message
    fun createProxyError(requestId: String, error: String): String {
        return JSONObject().apply {
            put("type", "proxy_error")
            put("requestId", requestId)
            put("error", error)
        }.toString()
    }

    // Create connect established message
    fun createConnectEstablished(requestId: String): String {
        return JSONObject().apply {
            put("type", "connect_established")
            put("requestId", requestId)
        }.toString()
    }

    // Create binary frame: requestId (36 bytes) + payload
    fun createBinaryFrame(requestId: String, data: ByteArray): ByteArray {
        val idBytes = requestId.padEnd(REQUEST_ID_LENGTH).toByteArray(Charsets.UTF_8)
        return idBytes + data
    }

    // Parse binary frame
    fun parseBinaryFrame(data: ByteArray): Pair<String, ByteArray> {
        val requestId = String(data, 0, REQUEST_ID_LENGTH, Charsets.UTF_8).trim()
        val payload = data.copyOfRange(REQUEST_ID_LENGTH, data.size)
        return Pair(requestId, payload)
    }
}

File: server/src/websocket/protocol.ts
// === Server → Phone Messages ===

export interface ProxyRequestMessage {
  type: 'proxy_request';
  requestId: string;
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: string;
}

export interface ConnectRequestMessage {
  type: 'connect_request';
  requestId: string;
  host: string;
  port: number;
}

export interface ConfigUpdateMessage {
  type: 'config_update';
  ports: number[];
}

export interface PingMessage {
  type: 'ping';
}

export type ServerToPhoneMessage =
  | ProxyRequestMessage
  | ConnectRequestMessage
  | ConfigUpdateMessage
  | PingMessage;

// === Phone → Server Messages ===

export interface RegisterMessage {
  type: 'register';
  deviceKey: string;
  deviceInfo: {
    name: string;
    carrier?: string;
    ip?: string;
  };
}

export interface ProxyResponseHeadersMessage {
  type: 'proxy_response_headers';
  requestId: string;
  statusCode: number;
  headers: Record<string, string>;
}

export interface ProxyResponseEndMessage {
  type: 'proxy_response_end';
  requestId: string;
}

export interface ProxyErrorMessage {
  type: 'proxy_error';
  requestId: string;
  error: string;
}

export interface ConnectEstablishedMessage {
  type: 'connect_established';
  requestId: string;
}

export interface PongMessage {
  type: 'pong';
}

export type PhoneToServerMessage =
  | RegisterMessage
  | ProxyResponseHeadersMessage
  | ProxyResponseEndMessage
  | ProxyErrorMessage
  | ConnectEstablishedMessage
  | PongMessage;

// === Binary Frame Helpers ===
// Binary frames: first 36 bytes = requestId (UUID string), remaining = payload

export const REQUEST_ID_LENGTH = 36;

export function createBinaryFrame(requestId: string, data: Buffer): Buffer {
  const idBuffer = Buffer.from(requestId.padEnd(REQUEST_ID_LENGTH));
  return Buffer.concat([idBuffer, data]);
}

export function parseBinaryFrame(data: Buffer): { requestId: string; payload: Buffer } {
  const requestId = data.subarray(0, REQUEST_ID_LENGTH).toString().trim();
  const payload = data.subarray(REQUEST_ID_LENGTH);
  return { requestId, payload };
}
